Align web terminal messages with the VM

srujzs opened the pull request:
- Adds better instructions for hot reload (if using the right flags), hot restart, quitting, clearing, and more. These were already being printed when using the VM, so this aligns with that.
- Adds an extra parameter for `CommandHelp` to `ResidentRunner` so `ResidentWebRunner` can pass a version of it that uses its separate logger and not `globals`. In order to support this, classes up the stack also provide a `Terminal`, `Platform`, and `OutputPreferences`.
- Fixes up use of `globals` from an earlier change to implement hot reload to use the logger instead. Same with `globals.platform`.
- Adds tests to check that only hot restart is printed when not using the extra front-end flags, and both hot restart and hot reload is printed when you are.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

srujzs committed:
Align web terminal messages with the VM

Adds better instructions for hot reload (if using the right flags),
hot restart, quitting, clearing, and more.

srujzs committed:
Merge remote-tracking branch 'upstream/master' into webterminallogs

srujzs committed:
Add param to ResidentRunner to accept a CommandHelp and add tests to check that output is correct

biggs0125 commented:
It'd be nice to share this code between the `ReloadHot` runner and our `ResidentWebRunner`... and even `ReloadCold` which is very similar. We're likely to just get more divergence over time if we have multiple copies of this code.

Can we just add 2 boolean getters to the `ResidentRunner`: `supportsReload` and `supportsDetach`? Then `printHelp` is just a single implementation on `ResidentRunner` and it hides `r` and `d` based on those flags. I think everything else is the same between all 3 configurations.

Eventually we can look into implementing `detach` on the web and then we can even delete `supportsDetach`.

srujzs committed:
Share implementation of printHelp

- Adds supportsDetach getter to determine whether to print
the command for detach.
- Adds extra parameter to printHelp to avoid printing hot
reload command when hot reload is implemented with hot restart.

srujzs commented:
I originally shied away from that as there exists a boolean for hot reload: `canHotReload`. This is always set to true for the web, even though hot reload is implemented as hot restart. This boolean also determines whether to do anything when the user types 'r', so I didn't want to start returning false for it.

However, I think adding an extra parameter to `printHelp` to tell it not to print the command for hot reload when not using the DDC library bundle format is sufficient here to maintain the existing behavior.

There was also some oddness around the VM using `globals` and the web using `_logger` to log things, but `ResidentRunner.logger` is the respective expected value anyways, so a shared implementation of `printHelp` is possible.

It looks like the VM only supports detach under certain conditions, so `supportsDetach` will likely stay.

srujzs committed:
Add missing new lines to test expectations

biggs0125 commented on packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:
```diff
@@ -105,6 +105,12 @@ class ResidentWebRunner extends ResidentRunner {
          stayResident: stayResident,
          machine: machine,
          devtoolsHandler: devtoolsHandler,
+         commandHelp: CommandHelp(
```
Do we still need this? I notice the logger might be different but I'm not clear how/why that would happen.

srujzs replied:
Unfortunately, yes. Some web tests are set up to provide a custom logger to the `ResidentWebRunner` (like the one in this PR) and then examine the output. This need to be piped to the `CommandHelp` that outputs the text.

Outside of tests, it looks like only `globals.logger` is ever passed to `ResidentWebRunner` through several other classes. One could probably clean-up all those classes and these tests so `logger` is no longer a constructor parameter, but I'm not sure what other issues they may come across in that refactor.

biggs0125 replied:
Thanks for confirming!

biggs0125 reviewed (approved):
Thanks for aligning these!

bkonyi commented on packages/flutter_tools/lib/src/resident_runner.dart:
```diff
@@ -1453,6 +1463,32 @@ abstract class ResidentRunner extends ResidentHandlers {
   }
 
+  @override
+  void printHelp({required bool details, bool reloadIsRestart = false}) {
```
Can we instead add `reloadIsRestart` as a getter on `ResidentRunner` that defaults to returning `false` and override it in the `ResidentWebRunner`? Then we don't need to add an additional parameter here and won't need to override the implementation.

Another option could be to just access the `debuggingOptions` directly here and perform the check from above (probably with an additional condition to ensure we're actually running against a web target) directly in this function:

```
debuggingOptions.buildInfo.ddcModuleFormat != DdcModuleFormat.ddc ||
debuggingOptions.buildInfo.canaryFeatures != true,
```

srujzs replied:
I didn't really like any option here when I thought about it, since `reloadIsRestart` is a weird getter to have since it only exists for legacy web behavior, and introducing web-specific logic into `ResidentRunner` also felt like a code smell. I think the getter is better than overriding the method and adding an extra parameter though, so I went with that.

srujzs committed:
Address review comments

srujzs commented on packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:
```diff
@@ -146,6 +152,10 @@ class ResidentWebRunner extends ResidentRunner {
 
   bool get _enableDwds => debuggingEnabled;
 
+  // TODO(srujzs): We should support this. Currently, it quits the app.
```
Done.

bkonyi reviewed (approved):
Thanks for the changes! LGTM!
